### PR TITLE
Faster SystemVM Template Builds

### PR DESCRIPTION
**Pull request by rohityadavcloud**
This introduces a new systemvmtemplate build mechanism that uses qemu-nbd, debootstrap and chroot to build systemvmtemplates. Compared to packer based builds which can take at least 30mins, with this PR the build time is now 5-10 mins.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: systemvm/image/scripts/configure_systemvm_services.sh [hunk 1]
```diff
@@ -50,9 +50,11 @@ function configure_cacerts() {
   CDIR=$(pwd)
   cd /tmp
   # Add LetsEncrypt ca-cert
-  wget https://letsencrypt.org/certs/lets-encrypt-x3-cross-signed.der
-  keytool -trustcacerts -keystore /etc/ssl/certs/java/cacerts -storepass changeit -noprompt -importcert -alias letsencryptauthorityx3cross -file lets-encrypt-x3-cross-signed.der
-  rm -f lets-encrypt-x3-cross-signed.der
+  wget https://letsencrypt.org/certs/isrgrootx1.der
+  wget https://letsencrypt.org/certs/lets-encrypt-r3.der
+  keytool -trustcacerts -keystore /etc/ssl/certs/java/cacerts -storepass changeit -noprompt -importcert -alias letsencryptauthorityx1 -file isrgrootx1.der
+  keytool -trustcacerts -keystore /etc/ssl/certs/java/cacerts -storepass changeit -noprompt -importcert -alias letsencryptauthorityr3 -file lets-encrypt-r3.der
```
Comment (rohityadavcloud) [inline systemvm/image/scripts/configure_systemvm_services.sh:56]: cc @Pearl1594 pl use/sync changes from this PR with your systemvmtemplate improvements for 4.16.1

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by rohityadavcloud** [comment]
@Pearl1594 quick question - now that we've moved to bundling all 3x templates (as default). Does this optimisation/build PR make sense? (the cons of this PR is that while it may be faster, it's (a) something new but not really, (b) requires sudo/root access to run nbd based build)

**Comment by Pearl1594** [comment]
Irrespective of the fact that we would bundle templates as part of the packages, we'd still need to build them, and this would improve the time taken to build the templates gravely. Also, I would assume there are certain users who leverage the template building tool to build custom templates. So yes, I think it would be beneficial

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@nvazquez conflicts fixed.
@Pearl1594 you may want to review/takeover this PR since your systemvm optimisation PR in. Check if we need to sync changes before/after merging the PR #5831

**Comment by nvazquez** [comment]
Thanks @rohityadavcloud

**Comment by nvazquez** [comment]
@blueorangutan package

File: systemvm/image/scripts/install_systemvm_packages.sh [hunk 2]
```diff
@@ -96,10 +90,9 @@ function install_packages() {
   # Install docker and containerd for CKS
   curl -fsSL https://download.docker.com/linux/debian/gpg | sudo apt-key add -
   apt-key fingerprint 0EBFCD88
-  add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"
+  echo "deb [arch=amd64] https://download.docker.com/linux/debian bullseye stable" > /etc/apt/sources.list.d/docker.list
   apt-get update
   ${apt_get} install containerd.io
-
   apt_clean
```
Comment (nvazquez) [inline systemvm/image/scripts/install_systemvm_packages.sh:96]: Hi @rohityadavcloud @Pearl1594 the script fails for me on Ubuntu 20.04 at this line:

````
/scripts/install_systemvm_packages.sh: line 96: apt_clean: command not found
+ fail 'install_systemvm_packages.sh failed'
+ clean_env
+ '[' /tmp/tmp.x7wCjvUAzz '!=' '' ']'
+ chroot /tmp/tmp.x7wCjvUAzz umount /proc /sys /dev /boot
+ '[' /tmp/tmp.x7wCjvUAzz '!=' '' ']'
+ umount /tmp/tmp.x7wCjvUAzz
+ sync
+ '[' /dev/nbd0 '!=' '' ']'
+ qemu-nbd -d /dev/nbd0
/dev/nbd0 disconnected
+ sync
+ '[' /tmp/tmp.x7wCjvUAzz '!=' '' ']'
+ rm -fr /tmp/tmp.x7wCjvUAzz
+ echo 'Failed: install_systemvm_packages.sh failed'
Failed: install_systemvm_packages.sh failed
+ exit 1
````

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

File: systemvm/image/scripts/install_systemvm_packages.sh (hunk 2 above)
Comment (nvazquez) [reply systemvm/image/scripts/install_systemvm_packages.sh:96]: After removing line 96 it suceeds (prepended time command):
````
SystemVM image generated: image.qcow2

real	5m37.860s
user	2m48.715s
sys	1m6.678s
````

**Comment by weizhouapache** [comment]
@rohityadavcloud 
great job !

@nvazquez @Pearl1594 
can you build the templates and kick off tests with new the templates ?

**Comment by rohityadavcloud** [comment]
Worth noting, there are two steps to generating the template; first the build script and then the export script. Unfortunately to use nbd, the build script must be run with `sudo` but the export script doesn't require sudo.

    sudo bash -x export.sh image.qcow2

File: systemvm/image/scripts/install_systemvm_packages.sh (hunk 2 above)
Comment (rohityadavcloud) [reply systemvm/image/scripts/install_systemvm_packages.sh:102]: Thanks @nvazquez for testing; indeed the build time looks fair (compared to packer which takes about 20-40mins)

**Comment by rohityadavcloud** [comment]
@Pearl1594 pl do advise if I need to make any changes, or if these smoketests were built/testing against this PR?

**Comment by Pearl1594** [comment]
Sure @rohityadavcloud . I made one change a few days ago due to a missing method - sorry.
Regarding smoke tests, I don't think it was run with the systemVM template built using this PR. I'll let @nvazquez confirm.

**Comment by nvazquez** [comment]
@rohityadavcloud @Pearl1594 I'm building the new system VM templates and will kick a new round of tests with them

**Comment by nvazquez** [comment]
Looks like there is an issue for KVM and Xen system VM templates, tried booting it up and fails. Moving it to the next milestone, still needs work
image

**Comment by rohityadavcloud** [comment]
Thanks @nvazquez yes let's revisit this for the next release. It may be too risky to consider this PR for 4.17 at this point.

**Comment by rohityadavcloud** [comment]
Closing this for now, may revisit this in future.

**Pull request closed without merging**